pkg/resource/deploy/step_generator.go

```diff
@@ -619,6 +614,122 @@ func (sg *stepGenerator) issueCheckErrors(new *resource.State, urn resource.URN,
 	return true
 }
 
+func (sg *stepGenerator) getResourceProvider(
+	urn resource.URN, custom bool, provider string, typ tokens.Type) (plugin.Provider, error) {
+
+	// If this is not a custom resource, then it has no provider by definition.
+	if !custom {
+		return nil, nil
+	}
+
+	// If this resource is a provider resource, use the plan's provider registry for its CRUD operations.
+	// Otherwise, resolve the the resource's provider reference.
+	if providers.IsProviderType(typ) {
+		return sg.plan.providers, nil
+	}
+
+	contract.Assert(provider != "")
+	ref, refErr := providers.ParseReference(provider)
+	if refErr != nil {
+		return nil, errors.Errorf("bad provider reference '%v' for resource '%v': %v", provider, urn, refErr)
+	}
+	p, ok := sg.plan.GetProvider(ref)
+	if !ok {
+		return nil, errors.Errorf("unknown provider '%v' for resource '%v'", ref, urn)
+	}
+	return p, nil
+}
+
+func (sg *stepGenerator) calculateDependentReplacements(root *resource.State) ([]*resource.State, error) {
+	// We need to compute the set of resources that may be replaced by a change to the resource under consideration.
+	// We do this by taking the complete set of transitive dependents on the resource under consideration and
+	// removing any resources that would not be replaced by changes to their dependencies. We determine whether or not
+	// a resource may be replaced by substituting unknowns for input properties that may change due to deletion of the
+	// resources their value depends on and calling the resource provider's `Diff` method.
+	//
+	// This is perhaps clearer when described by example. Consider the following dependency graph:
+	//
+	//       A
+	//     __|__
+	//     B   C
+	//     |  _|_
+	//     D  E F
+	//
+	// In this graph, all of B, C, D, E, and F transitively depend on A. It may be the case, however, that changes to
+	// the specific properties of any of those resources R that would occur if a resource on the path to A were deleted
+	// and recreated may not cause R to be replaced. For example, the edge from B to A may be a simple `dependsOn` edge
+	// such that a change to B does not actually influence any of B's input properties. In that case, neither B nor D
+	// would need to be deleted before A could be deleted.
```

Comment:
The `for example` is maybe too strong of a case.  It's worth calling out also here the more "normal" case where the edge is a dependencies via a property that does not force replacement, since that's the subtle case where we are creating a new semantic for `Diff`.  I think it's the right thing, but that's the piece I'd make as clear as possible in comments here if possible.